mmv1/third_party/terraform/services/container/resource_container_cluster.go.erb

```diff
@@ -3663,6 +3685,44 @@ func resourceContainerClusterUpdate(d *schema.ResourceData, meta interface{}) er
 		}
 	}
 
+	if d.HasChange("enable_k8s_beta_apis") {
+		log.Print("[INFO] Enable Kubernetes Beta APIs")
+		if v, ok := d.GetOk("enable_k8s_beta_apis"); ok {
+			name := containerClusterFullName(project, location, clusterName)
+			clusterGetCall := config.NewContainerClient(userAgent).Projects.Locations.Clusters.Get(name)
+			if config.UserProjectOverride {
+				clusterGetCall.Header().Add("X-Goog-User-Project", project)
+			}
+			// Fetch the cluster information to get the already enabled Beta APIs.
+			cluster, err := clusterGetCall.Do()
+			if err != nil {
+				return err
+			}
+
+			// To avoid an already enabled Beta APIs error, we need to deduplicate the requested APIs
```

Comment:
@coolkp just want to double-check that `DesiredK8sBetaApis` requires all listed APIs to not already be enabled?

My previous recommendation (to rely on API errors) was based on the understanding that we would send the full set of enabled APIs with each request, and that the endpoint would respond with an error if any enabled APIs were missing from the request (because they can't be disabled).

If we can only send APIs that have not been enabled yet - then we probably need to use a CustomizeDiff to conditionally force recreation of the resource if any APIs have been removed from the enabled set. See https://googlecloudplatform.github.io/magic-modules/best-practices/#forcenew for more details.

Reply:
Yeah it requires all enabled apis, if its missing will reject is as attempt to disable

Reply:
Sorry to bother you again. I am wondering if `DesiredK8sBetaApis` only accepts not-enabled APIs.
I ask because I am trying to use `DesiredK8sBetaApis` to include both an already enabled API (i.e. `authentication.k8s.io/v1beta1/selfsubjectreviews`) and a new dummy invalid API (i.e. `flowcontrol.apiserver.k8s.io/v1beta3/flowschema`).
However, the GKE API rejects the request with the error `Beta API "authentication.k8s.io/v1beta1/selfsubjectreviews" already specified.`.


I commented out the deduplication logic (l.4919-4923) as follows:

```go
func expandEnableK8sBetaApis(configured interface{}, enabledAPIs []string) *container.K8sBetaAPIConfig {
	l := configured.([]interface{})
	if len(l) == 0 || l[0] == nil {
		return nil
	}

	config := l[0].(map[string]interface{})
	result := &container.K8sBetaAPIConfig{}
	if v, ok := config["enabled_apis"]; ok {
		notEnabledAPIsSet := v.(*schema.Set)
		// for _, enabledAPI := range enabledAPIs {
		// 	if notEnabledAPIsSet.Contains(enabledAPI) {
		// 		notEnabledAPIsSet.Remove(enabledAPI)
		// 	}
		// }

		result.EnabledApis = tpgresource.ConvertStringSet(notEnabledAPIsSet)
	}

	return result
}
```

I built the provider manually and create the cluster using the following Terraform configuration:

```HCL
data "google_container_engine_versions" "asia_northeast1" {
  location = "asia-northeast1"
}

resource "google_container_cluster" "primary" {
  name     = "test-cluster"
  location = "asia-northeast1"

  min_master_version = data.google_container_engine_versions.asia_northeast1.release_channel_latest_version["STABLE"]
  enable_k8s_beta_apis {
    enabled_apis = [
      "authentication.k8s.io/v1beta1/selfsubjectreviews",
    ]
  }

  node_pool {
    name               = "default-pool"
    initial_node_count = 1
    management {
      auto_repair  = false
      auto_upgrade = false
    }
  }
}
```

Add dummy invalid API to `flowcontrol.apiserver.k8s.io/v1beta3/flowschema` field and apply the configuration:

```HCL
data "google_container_engine_versions" "asia_northeast1" {
  location = "asia-northeast1"
}

resource "google_container_cluster" "primary" {
  name     = "test-cluster"
  location = "asia-northeast1"

  min_master_version = data.google_container_engine_versions.asia_northeast1.release_channel_latest_version["STABLE"]
  enable_k8s_beta_apis {
    enabled_apis = [
      "authentication.k8s.io/v1beta1/selfsubjectreviews",
      "flowcontrol.apiserver.k8s.io/v1beta3/flowschema", # This is invalid API
    ]
  }

  node_pool {
    name               = "default-pool"
    initial_node_count = 1
    management {
      auto_repair  = false
      auto_upgrade = false
    }
  }
}
```

Then got the following error:

```console
❯ TF_CLI_CONFIG_FILE="/Users/tsubasanagasawa/tf-dev-override.tfrc" terraform apply

...

Terraform will perform the following actions:

  # google_container_cluster.primary will be updated in-place
  ~ resource "google_container_cluster" "primary" {
        id                          = "projects/*******/locations/asia-northeast1/clusters/test-cluster"
        name                        = "test-cluster"
        # (26 unchanged attributes hidden)

      ~ enable_k8s_beta_apis {
          ~ enabled_apis = [
              + "flowcontrol.apiserver.k8s.io/v1beta3/flowschema",
                # (1 unchanged element hidden)
            ]
        }

        # (17 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

google_container_cluster.primary: Modifying... [id=projects/*****/locations/asia-northeast1/clusters/test-cluster]

Error: googleapi: Error 400: Beta API "authentication.k8s.io/v1beta1/selfsubjectreviews" already specified.
Details:
[
  {
    "@type": "type.googleapis.com/google.rpc.RequestInfo",
    "requestId": "0xd2fd528fadf2358a"
  }
]
, badRequest

  with google_container_cluster.primary,
  on gke.tf line 1, in resource "google_container_cluster" "primary":
   1: resource "google_container_cluster" "primary"

Releasing state lock. This may take a few moments...
```

Reply:
@toVersus thanks for trying it out & @coolkp for double-checking the behavior. It looks like the API indeed only takes the APIs that should be enabled, so deduping is the correct behavior :-) We should also set up a CustomizeDiff to conditionally force recreation of the cluster if an enabled API is removed, as described in https://googlecloudplatform.github.io/magic-modules/best-practices/#forcenew.

Reply:
Thanks for double-checking the behavior! I'm glad that I was right in my thinking 😄
I added a CustomizeDiff in f12421f.